web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx

```diff
@@ -0,0 +1,175 @@
+  }
+
+  const getCurrentClusterPinnedResources = useCallback(
+    () => getClusterPinnedResources(clusterId),
+    [clusterId, getClusterPinnedResources]
+  );
+  const updateCurrentClusterPinnedResources = useCallback(
+    (pinnedResources: string[]) =>
+      updateClusterPinnedResources(clusterId, pinnedResources),
+    [clusterId, updateClusterPinnedResources]
+  );
+
+  return (
+    <SharedUnifiedResources
+      params={params}
+      updateUnifiedResourcesPreferences={preferences => {
+        updatePreferences({ unifiedResourcePreferences: preferences });
+      }}
+      availableKinds={['app', 'db', 'windows_desktop', 'kube_cluster', 'node']}
+      Header={pinAllButton => (
+        <>
+          <FeatureHeader
+            css={`
+              border-bottom: none;
+            `}
+            mb={1}
+            alignItems="center"
+            justifyContent="space-between"
+          >
+            <FeatureHeaderTitle>Resources</FeatureHeaderTitle>
+            <Flex alignItems="center">
+              <AgentButtonAdd
+                agent={SearchResource.UNIFIED_RESOURCE}
+                beginsWithVowel={false}
+                isLeafCluster={isLeafCluster}
+                canCreate={canCreate}
+              />
+            </Flex>
+          </FeatureHeader>
+          <Flex alignItems="center" justifyContent="space-between">
+            <SearchPanel
+              params={params}
+              pathname={pathname}
+              replaceHistory={replaceHistory}
+              setParams={setParams}
+            />
+            {pinAllButton}
+          </Flex>
+        </>
+      )}
+      setParams={newParams => {
+        setParams(newParams);
+        replaceHistory(
+          encodeUrlQueryParams(
+            pathname,
+            newParams.search,
+            newParams.sort,
+            newParams.kinds,
+            !!newParams.query /* isAdvancedSearch */,
+            newParams.pinnedOnly
+          )
+        );
+      }}
+      key={clusterId} // when the current cluster changes, remount the component
+      getClusterPinnedResources={getCurrentClusterPinnedResources}
+      updateClusterPinnedResources={updateCurrentClusterPinnedResources}
+      pinningNotSupported={pinningNotSupported}
+      onLabelClick={onLabelClick}
+      EmptySearchResults={
+        <Empty
+          clusterId={clusterId}
+          canCreate={canCreate && !isLeafCluster}
+          emptyStateInfo={emptyStateInfo}
+        />
+      }
+      fetchFunc={useCallback(
+        async (params, signal) => {
+          const resp = await teleCtx.resourceService.fetchUnifiedResources(
+            clusterId,
+            params,
+            signal
+          );
+
+          return {
+            startKey: resp.startKey,
+            agents: resp.agents.map(resource => ({
+              resource,
+              ui: {
+                ActionButton: <ResourceActionButton resource={resource} />,
+              },
```

Comment:
I'm not sure the action button should be returned in the fetch func. Perhaps I'm missing a step but this button wont be 'static' in access requests and requires state from access requests (if its added to the request or not).

Reply:
This makes the API a bit more complex, but I think you are right. Fetching and mapping items to the rendered form should be separated.